streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerializer.java

```diff
@@ -0,0 +1,38 @@
+package org.apache.kafka.streams.kstream.internals.foreignkeyjoin;
+
+import org.apache.kafka.common.serialization.Serializer;
+
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+public class SubscriptionResponseWrapperSerializer<V> implements Serializer<SubscriptionResponseWrapper<V>> {
+    private final Serializer<V> serializer;
+
+    public SubscriptionResponseWrapperSerializer(Serializer<V> serializer) {
+        this.serializer = serializer;
+    }
+
+    @Override
+    public void configure(Map configs, boolean isKey) {
+        //Do nothing
+    }
+
+    @Override
+    public byte[] serialize(String topic, SubscriptionResponseWrapper<V> data) {
+        //{16-bytes Hash}{n-bytes serialized data}
+        byte[] serializedData = serializer.serialize(null, data.getForeignValue());
```

Comment:
Why is the topic passed as null? It causes issues with GenericRecord AVRO serializer, since it tries to register schemas under "null-value" subject, and the schema registry responds with "version not compatible" error

Reply:
The issue is actually with the Confluent implementation of the SerDe, as they incorrectly attempt to register when null topics are passed in. Read https://github.com/confluentinc/schema-registry/issues/1061 for more details. That being said, it has been extremely quiet in that git repo, I am not sure how much effort Confluent puts into supporting work on that product.

Reply:
If this does not gets fixed either way, this PR will be unusable for most of the practical use cases. What is the downside of passing the topic name to the serializer? I tried it, and it seemed to work as expected.
Is there a workaround if confluentinc/schema-registry#1061 is not fixed?

Reply:
I think the main issue would be the large amount of internal topic schemas registered to the schema registry. This, combined with any breaking changes to the schema (due to normal business requirement changes) would make it such that you are now needing to manually delete schema registry entries made to internal topics. This is a workflow that I do not believe was ever intended to be done with the Confluent Serde.

As it stands right now, there are allegedly other functionalities that require null serialization ("There are several places in Streams where we need to serialize a value for purposes other than sending it to a topic (KTableSuppressProcessor comes to mind), and using `null` for the topic is the convention we have."). These too will not work with the confluent Serde.

If they do not fix it, then the next best thing to do would be wrap it in your own implementation and intercept null-topic values to avoid registration.  I do not see why it wouldn't be fixed since the current behaviour of registering "null-topic" is fundamentally useless. 

Anyways, with all that being said, for this particular line I can certainly pass in the topic since it's fairly well-defined. For other parts, such as https://github.com/apache/kafka/blob/18a16b8eda3fb3e26d3fd6e7c209ceb983814a3e/streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/KTableRepartitionerProcessorSupplier.java#L67, there is no solution using the current Confluent Serde.

Reply:
Confluent serde needs a schema id, and looks like it is not stored in GenericData.Record instance - it may not be trivial to fix  confluentinc/schema-registry#1061...

Reply:
That is a fair point.

@mjsax @vvcephei @guozhangwang: It may not be as easy to get the hash value as we thought if we rely on the value serde, namely because of the interactions with the schema registry to serialize the data. Though this is technically a confluent issue, it may be sufficient to put a blocker on this PR since, as Adaniline demonstrates, many users of this use the confluent serde. Any thoughts on a way forward are welcome.

Reply:
Hmm. This does seem like an impediment. The Serializer interface doesn't specifically state whether the topic is nullable or non-nullable. What it does say is `@param topic topic associated with data`. Getting fully into armchair lawyer mode, it seems like this statement implies non-nullability, since `null` is not a valid topic name. If it could be null, it should say something like "the topic associated with the data, or null if there is none."

Also, backing up to a higher level, the Serializer interface is specifically for serializing data for use with Kafka topics. So, it's probably not appropriate to use it for general serialization. If we want to do that, maybe we should add a new interface that doesn't imply we're sending data to a topic.

Then again, this is in support of an optimization. Perhaps we should just drop the optimization for now and go back to storing only the foreign key reference for the correctness check. For bonus points, we could design a message and storage format that leaves the door open for future optimizations like this, without an awkward upgrade procedure.

WDYT, @guozhangwang ?

Reply:
@bellemare @vvcephei On the high-level, I think this should be fixed at Confluent SR, rather than letting an AK feature blocked on it -- i.e. users should still be able to pass in not-null topic names while non-schema required typed data will not be registered.

Reading on the source code of avro-serializer, I think the root cause is that today we treat all primitive typed data as a simple avro-schema as well when serializing, hence still registering it. Details are here:

https://github.com/confluentinc/schema-registry/blob/master/avro-serializer/src/main/java/io/confluent/kafka/serializers/KafkaAvroSerializer.java#L52

and 

https://github.com/confluentinc/schema-registry/blob/8f727d74da65a9be4f7fd82fd8a2498ad48c85ec/avro-serializer/src/main/java/io/confluent/kafka/serializers/AvroSchemaUtils.java#L30

We could, instead, fix SR to let it just use primitive serdes directly for primitive types. In this way users can still provide topic names.

Reply:
FWIW @guozhangwang , as a user of the SR/serdes, I would consider "fix SR to let it just use primitive serdes directly for primitive type" to be a regression. Registering primitive schemas with the schema registry is a feature, because it allows other applications (support tooling, etc.) to understand the data in a topic without any context whatsoever.

Reply:
I think to answer the question of if Confluent needs to change their SR and serde, or if we need to take a different approach requires us to address John's first paragraph: "The Serializer interface doesn't specifically state whether the topic is nullable or non-nullable."

If the topic must be non-nullable, then the current approach with the hash code wont work, since the output topic is unknown by the KTableRepartitionerProcessorSupplier.java. If the topic must be nullable (and still provide a serialized output), then the confluent serde/SR needs to change to support this. This isn't clear from the contract and so it probably does need to be resolved.

Lastly, I do agree with pgwhalen on the usage of primitive schemas in support tooling. We use this functionality all the time with schema registry lookups, as opposed to having our team guess as to what it may have been serialized in. 

One last thought - for the confluent serde, if the topic is null and the type is a primitive, we can avoid registration and do as @guozhangwang outlined. If the type is not a primitive, it can throw an exception (as it effectively does now, upon trying to register to topic-null). This does seem a bit hacky and perhaps a bit complicated, but it may work.

Reply:
Can the solution be to have a separate interface to calculate data hash (HashSupplier, etc), and have Confluent Avro Serializer implement it? Looks like it is a separate concern from kafka data serialization
This way KTableRepartitionerProcessorSupplier could do something like
```java
if (valueSerializer instanceof HashSupplier) {
   currentHash = ((HashSupplier) valueSerializer).hash128(change.newValue);
} else {
   currentHash = Murmur3.hash128(valueSerializer.serialize(null, change.newValue)));
}
```

Reply:
> because it allows other applications (support tooling, etc.) to understand the data in a topic without any context whatsoever.

@pgwhalen I see. That's a valid point. After a second thought I feel that instead of all / never (i.e. always register a simple avro schema, v.s. never register for primitive types) we should give users options to choose from the two.

> Can the solution be to have a separate interface to calculate data hash (HashSupplier, etc), and have Confluent Avro Serializer implement it?

@adaniline-traderev That should also work, in fact, we can just let Streams to use a special serde other than whatever registered serde for this repartition topic's value, instead of asking the registered serde to provide a hashSerde function: note that, the current serde precedence is, from high to low:

1) User overridden value at the DSL.
2) Streams library hard-coded value internally (think: LongSerde for `count`).
3) Default registered serde value.

But I'm a bit concerned about adding more special handling of serdes at the Streams side, i.e. add more scenarios of case 2) above: today Streams try to reason about the resulted stream key/value type out of an operator in a best effort and try to use the serde correspondingly event if user does not specify any overrides. The goal is to try to be as adherent to case 1) as possible. Although we do have special hard-coded serdes as case 2), e.g. in `count` operator, we hard-coded the serde as LongSerde unless user overrides it, we still want to keep such cases as small as possible, since having Streams to special-handle, e.g. a `KTableRepartitionerProcessor` opens the door for more complexity, as this handling logic would be scattered across the classes and error-prone: note that besides the sink node, at the source node of the downstream sub-topology, Streams then should also remember to use a special serde instead of the registered one to deserialize it as well.

That being said, if people feels it is worth to add this complexity in Streams for good reasons, I can be convinced as well

----------------

Philosophically, I'd still argue that, the AK code should not yield to a single vendor's dependency when considering its logic (disclaimer: I'm currently employed at Confluent). We should consider what's the best approach for AK Streams here, and then if it affects the eco-system dependency we should fix it on the other side.

Reply:
@guozhangwang I don't disagree with your stand on AK first. I do, however, agree with John that the contract is not quite clear on if topic should always be non-null or if topic can be nullable. This is pivotal for determining the way forward. I believe that a precedent has been set by the suppress functionality ( https://github.com/apache/kafka/blob/0846f590dac5d3893d41e262d67ee8a228dcc332/streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java#L95 ), and if we wish to be consistent we should make clear that serializers need to handle null topics (though that handling could simply be throwing an exception.. as is effectively the case with the Confluent Serde / SR ).

Reply:
For this aspect, my personal take is Streams should pass in the topic name whenever it knows -- for the example you brought up, currently it's because the operator does not know the topic name I believe, for which we could refactor internal code to fix it (would like @vvcephei to chime in here whether that is really doable) -- and not relying on any assumptions of the serdes itself, whether the topic is nullable or not.

Reply:
Hah! Well, now I'm really in the middle of it ;) 

It's worth noting that there's plenty of precedent on both sides of the issue. I don't think the suppress implementation constitutes a contract. It could just as easily be argued that L95 there is a bug. Actually, I still feel the way I did toward the beginning of the thread, that it makes more sense to say it's non-nullable (given the javadoc and the origin/ownership of the interface in `kafka-clients`).

I'm happy to fix that bug in suppress, since we actually do send the serialized result there to a changelog topic, we should just provide the CL topic name to the serializer.

This case is more of a bind because I'm not aware of anywhere else in Streams where we legitimately serialize something we have no intention of sending to any topic.

Reply:
I think for suppress() it's a bug -- and it seems to break SR integration: on restore the schema needs to be fetches from the SR if AVRO is used, but with `null` topic on serialization the schema would never have be registered correctly and restore would fail with schema not found error.

For passing in `null` in general, I think it should be allowed, but I also think it's not really necessary. We can pass in the repartition topic name here, too.

With regard to the concern about writing something different into the topic compared to a registered schema: We do this already for windowed changelog topics (note that only the user-data-key is serialized as AVRO for this case, and we add one or two additional longs and maybe a unique counter before writing into the changelog). If we believe this is an issue, we can fix it, but I would not block this PR for it. It's orthogonal and if we think we need to fix it, we need to fix it for all topics. However, it was not an issue in practice so far, thus, my personal take it, that we should move forward with this PR and pass in the repartition topic name.

Reply:
@mjsax I can certainly pass in the repartition topic name for the serializer without any issue. At that point, it's up to the Confluent serde users to complain about their SR being populated with internal topic schemas (an issue we have run into at my company), though it's not a deal breaker. That being said, I am not aware of how to obtain the destination topic in https://github.com/apache/kafka/blob/18a16b8eda3fb3e26d3fd6e7c209ceb983814a3e/streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/KTableRepartitionerProcessorSupplier.java#L67. This processor creates the hash code for the current event from the value serde (ie: confluent's), but has no knowledge of the downstream repartition topic, and thus cannot provide a topic. It is my understanding that this would require information not available to the processor, and so will not work as currently structured. This is the real issue that hasn't been adequately addressed yet, aside from the suggestion from John to drop the hash code optimization. This would skip the whole issue, though it may be sub-optimal in terms of developer experience.

Reply:
At the risk of being annoying, the processor _could_ just make up a topic name to use. This would also fix your problem.

Right now, I think we have three viable solutions (in order of my preference):
1. generate a topic name for the processor to pass to the serializer
2. pass null to the serializer, forcing Confluent's SR and any other Serializer/SR to handle nulls gracefully.
3. drop the optimization

Option 2 is a little unfriendly to Confluent SR (and other Serializer) implementations, but pragmatically, we _are_ upstream, so we have this leverage we can apply to force them to change.

Reply:
Further in support of making up a topic name to give the serializer, this is what we're currently doing with state stores, when change-logging is disabled. We unconditionally generate the topic name to pass, given only the appId and storeName:

```java
    public static String storeChangelogTopic(final String applicationId,
                                             final String storeName) {
        return applicationId + "-" + storeName + STATE_CHANGELOG_TOPIC_SUFFIX;
    }
```

When change-logging is disabled for a store, there is no such topic; i.e., it's a made-up topic name. It's not immediately clear how this is different from something like: `applicationId + "-" + processorName + "-recordhash"`

Reply:
(tangential: I've just submitted https://github.com/apache/kafka/pull/6602 to fix the suppression serde handling. Please feel free to review it!)

Reply:
@vvcephei John, I do indeed like that option of `applicationId + "-" + processorName + "-recordhash"` because of the existing precedence with the made-up changelog topic. I believe a decision like this would further align us with the statement "topic should always be non-null in serialize/deserialize".

So it comes down to 3 options:

1) Punt any question on the SerDe topic nullability by removing the optimization.
2) Require that SerDe topic parameter be nullable by setting topic to null as I have already done in this PR.
3) Require that SerDe topic parameter be non-nullable by creating a fake topic name.

In order of preference, I believe 3 is most reasonable, followed by 1. I think 2 is actually fairly reasonable, but least likely to provide a productive way forward because of the existing ambiguous contract. Nullability would require that the loosest form of String be used.

If no one else objects, I'll go about using a `applicationId + "-" + processorName + "-recordhash"` and post an update shortly.

Reply:
Small change. Wasn't able to figure out a clean way to get processorName, but since it's a dummy anyways, I went with context().topic() instead of processorName

Reply:
Back to the SR / Avro-serde issue we discussed above: after thinking about that a bit more, I'm wondering if we should just fix it by hard coding a serde for the hash value.

Since our current serde precedence is:

1) User specified serde at the DSL control object.
2) Streams hard-coded serde on some special operator (e.g. at `count` operator we hard-code LongSerde).
3) Globally registered serde via config.

For the hashvalue, we can treat it as case 2) above and hard-code a serde internally that override the registered serde. And for this foreign key join operation we would not expose control objects for serde overrides case 1), hence this way we should be safe. WDYT?

Reply:
From my understanding, the design was to first serializer the data, and afterwards compute a hashvalue based on the `byte[]` array. However, for the first step, we don't know the type and thus cannot hard code a serializer.

Is my understanding incorrect?

Reply:
This is my understanding as well, which is why I do not see another way forward at the moment.

Reply:
Ah okay, I think I was the culprit for the misunderstanding here: I was thinking that we will 1) compute the hash code first, and then 2) serialize that hash code, and hence the issue is that SR also register for a primitive type which I thought should be addressed at Streams.

Now I realize that we are 1) serializing first, and then 2) compute the hash value based on the bytes. This actually makes much more sense since we cannot guarantee all typed objects have a consistent implementation of the `hashCode` interface. In this way though, the bottom line is that we are calling the serializer for getting the bytes, but we are not actually sending the bytes over the wire. And there are serializers like Confluent AvroSerde out there which tries to remember which topics are serialized with which schema, which will then break since Streams are not actually using the serialized bytes actually.

Given that scenario (hopefully I got it right this time :P), I think although it is indeed fixable on Confluent AvroSerde to not register schema given `null` topic name, it is still only one serde and we cannot tell there are no other serdes in the wild that relies on the passed in topic name parameters to bookkeep some mappings, and therefore I think we still need to consider this issue at the Streams layer. On this regard my preference is that, at the moment, we just pass in the actual topic name than using a consistent dummy topic name, and my reasoning is that repartition topics are really internal ones specific to the streams app itself, and not supposed to be exposed outside the app (though for changelog topics there may be some exceptions like other apps / consumers may want to read that topic directly, for repartition topics like this one I do feel that they should be really `transient` and not be leveraged by any external clients), and hence protecting such mis-usages sounds like an overkill to me.

In the long run, we should consider adding extra APIs in the general Serde interface which does not take extra parameters (for this purpose I think neither `topic` name or `isKey` boolean should be included) but just object -> bytes and vice versa, to leave to space for any serdes have semantical business logic around those fields. At that time we can then be on the safer side by choose the new APIs.

Reply:
@guozhangwang - Sounds very reasonable to me. My question is, do I now simply use `context().topic()`? I believe that is correct, since it'll be the topic that we're sourcing the data from, but I just want to verify.

Reply:
I discussed this with some other people, and somebody mentioned, that for the value we serialize, this value is actually also store in RocksDB (input `KTable`). We also know, that the corresponding `byte[]` are written into the store changelog topic. Hence, instead of using the repartition topic, using the changelog topic should be a better option, as it does not leak anything into SR (or whatever other Serdes might do with the topic name).

Even if there is not changelog topic for the input KTable (we do some optimizations and sometimes don't create one (eg, the store might actual be a logical view and is not materialized). But even for this case, using the changelog topic name seems to be save.

Reply:
`context().topic()` gives the repartition topic name in the serializer, which is what I want. In the processor sections, where I use null, `context().topic()` gives me the input-topic name for the KTable... which is also fine, since the serializer will check against the input topic schema, which must be valid by definition of the data being within the topic...  so I suspect this issue can be laid to rest, in line with adaniline-traderev's suggestion. 

This removes any requirement for the upstream serializer to have to do special work for null values.

Reply:
@mjsax I'm not sure I can fully follow the suggestion of using changelog topic v.s. the repartition topic here: are you suggesting to do it universally or just for this case? If it is the latter case, I felt it a bit awkward due to inconsistency with other source `KTable` cases where we will just follow the `SourceNode / RecordDeserializer` path to deserialize using the source topic; it if it the first case, that also has some drawbacks since with today's topology generation not all source `KTable`s will need to be materialized to a store and hence not necessary having a changelog topic.

I still feel that using the source topic name (and i.e. in this case, the repartition topic) admittedly exposed to SR but is philosophically the right thing to do, and we should consider fixing it on serde APIs in the future. WDYT

Reply:
@guozhangwang I was just talking about the foreign-key case (not sure why you thought it might be anything else?). My understanding is the following: The contract is that we should pass a topic name into the serializer of which we want to write the data into. This contract breaks if we pass in the repartition topic name, because we write something different into the repartition topic.

You are right that the changelog topic might not exist, however, my personal take is, that registering for a non-existing topic, is a smaller violation of the contract that passing in the "wrong" repartition topic name. Note, that the changelog topic name is conceptually the "right" topic name. However, this case would not happen very often anyway (compare examples below).

Your comment trigger one more thought: the optimization framework could actually check for different cases, and if there is an upstream topic (either changelog or source topic that has the same schema), we could actually use this name.

Some examples (does not cover all cases):

```
builder.table("table-topic").foreignKeyJoin(...)
```
For this case we need to materialize the base table (that is also the join-table), and the schema is registered on `table-topic` already, so we can pass in `table-topic` to avoid leaking anything.

```
builder.table("table-topic").filter(...).foreignKeyJoin(...)
```
For this case we materialize the derived table from the filter() and we get a proper `filter-changelog-topic` and we can pass this one.

```
builder.stream("stream-topic").groupBy().aggregate().foreignKeyJoin(...)
```
For this case, the agg result KTable is materialized and we can pass the `agg-changelog-topic` as name.

```
builder.stream("stream-topic").groupBy().aggregate().filter().foreignKeyJoin(...)
```
For this case, the agg result KTable is materialized and we can pass the `agg-changelog-topic` as name, because the filter() does not change the schema. Thus, even if the join-input KTable is not materialized, we can avoid to leak anything by "borrowing" the upstream changelog topic name of the filter input KTable.

```
builder.table("table-topic").mapValues(...).foreignKeyJoin(...)
```
For this case, we need to materialize the result of `mapValues()` and get a proper changelog topic for the join-input table.

```
builder.table("table-topic", Materialized.as("foo")).mapValues(...).foreignKeyJoin(...)
```
This might be a weird case, for which the base table is materialized, while the input join-table would not be materialized, and also the type changes via mapValues(). Hence, the `table-topic` schema is not the same as the join schema and we also don't have a changelog topic for the join-input KTable. We still use the changelog-topic name of the non-existent changelog topic (of the mapValues() result KTable).

As you can see, we can cover a large scope of cases for which we don't leak anything and can always use a topic name that contains data corresponding to the schema. Does this explain my thoughts?

Reply:
I understand your reasoning now, but still I felt Streams should not fix it trying to piggy-back on another topic that happens to be of the same schema that this serde is used for; or rather, I'd prefer to use a non-exist dummy topic than an existing topic if we do not like repartition topics (again, I agree that repartition topic is not ideal, since we are, in fact, not sending the bytes serialized in that way to the topic).